cluster/src/main/java/org/apache/iotdb/cluster/server/member/RaftMember.java

```diff
@@ -384,7 +385,89 @@ public void appendEntry(AppendEntryRequest request, AsyncMethodCallback resultHa
 
   @Override
   public void appendEntries(AppendEntriesRequest request, AsyncMethodCallback resultHandler) {
-    //TODO-Cluster#354: implement
+    logger.debug("{} received an AppendEntriesRequest", name);
+
+    // the term checked here is that of the leader, not that of the log
+    if (!checkRequestTerm(request, resultHandler)) {
+      return;
+    }
+
+    try {
+      long response = 0;
+      List<Log> logs = new ArrayList<>();
+      for (ByteBuffer buffer : request.getEntries()) {
+        Log log = LogParser.getINSTANCE().parse(buffer);
+        logs.add(log);
+      }
+
+      response = appendEntries(logs);
+      resultHandler.onComplete(response);
+      logger.debug("{} AppendEntriesRequest of log size {} completed", name,
+          request.getEntries().size());
+    } catch (UnknownLogTypeException e) {
+      resultHandler.onError(e);
+    }
+  }
+
+  /**
+   * Find the local previous log of "log". If such log is found, discard all local logs behind it
+   * and append "log" to it. Otherwise report a log mismatch.
+   *
+   * @param logs
+   * @return Response.RESPONSE_AGREE when the log is successfully appended or Response
+   * .RESPONSE_LOG_MISMATCH if the previous log of "log" is not found.
+   */
+  private long appendEntries(List<Log> logs) {
```

Comment:
I've found that this way of appendEntries has a high concurrency bug, this is the fix [document](https://shimo.im/docs/JWhvXw8vKyqvt6X8). Please refer to function `appendEntry`'s implementation.

Reply:
I guess LebronAl is talking about the chance that the logs are disjoint with the local last log, but as `appendEntries` is only used in catch-up, I think the possibility of that case is relatively low.

Still, it would be better to check the index of the first coming log and the last local log, which I previously hoped to be done inside the LogManager. As you can see, other users may not bother to do the checks (the may not know the checks actually), so any necessary checks are better to be done within underlined classes.

Reply:
Actually,I used to think that the interface raftLogManager exposed to followers to call append logs was `maybeAppend`, and in that function we can handle this check, but now appendEntry's RPC doesn't have lastLogTerm, lastLogIndex, leaderCommit. So i have to use `append `.
So now we can either add some fields to the appendEntry rpc to support the use of maybeAppend, or we can do this check inside `append`. What's your opinion?

Reply:
In my view, `maybeAppend ` will make catchUp more fault-tolerant.When catching up in batch, even if the index of the first log sent is smaller than follower's lastLogIndex, some previous logs can be removed. Otherwise, if the index of the first Log is smaller than  follower's commitindex, such a batch of logs cannot be appended.